### ISSUE TITLE
Inconsistent tf.losses.mean_squared_error on cpu vs gpu

### ISSUE THREAD

**Issue by cheyennee** [tensorflow/tensorflow#57172]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

TF2.4

### Custom Code

Yes

### Current Behaviour?

```shell
The result on CPU should be equal to the result on GPU.
```

### Standalone code to reproduce the issue

```shell
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_tensor = tf.saturate_cast(tf.random.uniform([2, 3], minval=-256, maxval=257, dtype=tf.int64), dtype=tf.int64)
      arg_0 = tf.identity(arg_0_tensor)
      arg_1_tensor = tf.random.uniform([2, 3], dtype=tf.float16)
      arg_1 = tf.identity(arg_1_tensor)
      results["res_cpu"] = tf.losses.mean_squared_error(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = tf.identity(arg_0_tensor)
      arg_0 = tf.cast(arg_0, tf.int64)
      arg_1 = tf.identity(arg_1_tensor)
      arg_1 = tf.cast(arg_1, tf.float16)
      results["res_gpu"] = tf.losses.mean_squared_error(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
'''
{'res_cpu': <tf.Tensor: shape=(2,), dtype=float16, numpy=array([inf, inf], dtype=float16)>, 'res_gpu': <tf.Tensor: shape=(2,), dtype=float16, numpy=array([35650., 25060.], dtype=float16)>}
'''
The result on CPU outputs inf array, while the result on GPU outputs float array.
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@cheyennee,

**tf.random.uniform** provides the output random values from a uniform distribution. The generated values follow a uniform distribution in the range [minval, maxval].  It is intended for the CPU and GPU to return different results if you pass random values as the input. 

- Use `tf.keras.utils.set_random_seed`(some_seed) control all random behavior is needed

- Optimize the algorithm as much as possible to mitigate the impact of calculation accuracy. That is, let the algorithm not produce or accumulate small errors in accuracy. There are many adjustable keys, not only the optimizer or the optimizer's parameters, but also other stuff such as bacth_size, gradient_clip, gradient_norm etc.

- Use `tf.config.experimental.enable_op_determinism` if you want to get completely consistent reproducible results and accept its side effects(deceleration).
Thank you!

**Comment by tilakrayal** [comment]
Also, setting the seeds is not enough. If you are working on a small dataset you have to switch from GPU to CPU, as the first one has a non-deterministic order of computations.

Moreover, you should get rid of any parallelism in feeding data, as that can also cause problems. Thank you!

**Comment by cheyennee** [comment]
@tilakrayal  Thanks for elaborating on tf.random.uniform usage. But in my code, I just use tf.random.uniform to output two random values. Then I feed these two random values  to CPU and feed same values  to GPU.  That is to say, CPU and GPU get same inputs. What I focus on here is that the CPU and GPU produce different outputs with the same inputs.

**Comment by tilakrayal** [comment]
@cheyennee,
I tried to execute the code multiple times on both cpu and gpu and I was able to fetch the different results for every execution on both [CPU](https://colab.research.google.com/gist/tilakrayal/04920f203454aa08761518266525000b/cpu.ipynb) and [GPU](https://colab.research.google.com/gist/tilakrayal/323f89e2ac22cf8b9c8d34829c814545/gpu.ipynb). Kindly find the gist.
As mentioned tf.random.uniform provides the output random values from a uniform distribution which are resulting in the output to have the inconsistency on both CPU and GPU. Thank you!
